pkg/eval/vals/aliased_types.go

```diff
@@ -7,8 +7,37 @@ import (
 	"github.com/xiaq/persistent/vector"
 )
 
-// File is an alias for *os.File.
-type File = *os.File
+// File wraps *os.File so that Elvish can treat the object as a read-only psuedo-map.
+type File struct {
+	OsFile *os.File `elvish:"hidden"`
+	// IoMode causes `$f[io-mode]` to return "ro" if the file is read-only, "rw" if it is
+	// read-write, and "wo" if it is write-only.
+	IoMode string
```

Comment:
In general, given an `os.File`, it is not possible to know the mode it was opened with; this makes the `File` type here only suitable for wraping `os.File` created in the Elvish runtime and limits interoperability. Better to just leave out this field.

Reply:
True, but at the moment we don't wrap any `os.File` object that we didn't create. We might want to wrap the equivalent of `os.Stdin`, et.al. in the future where we do not know the I/O mode. So rather than remove the field, since it is potentially useful for diagnostic purposes, I'll document that it could be `"??"` if Elvish doesn't know the I/O mode. Too, if we do ever expose the `os.Std*` objects as "file" objects that can be directly manipulated we can reasonably default the `IoMode` to "ro" if we're wrapping `os.Stdin`, otherwise "wo".

Reply:
That can certainly work, but after thinking more about this matter I think it's best to not wrap the `*os.File` type at all. Instead, to support introspection on `*os.File` objects, add it to the type switch in the `Index` function. My experience over working on the value system of Elvish suggests that it's best to stick to the types offered by Go as much as possible.

For the purpose of this PR, I recommend that you just leave out all the introspection-related changes (i.e. all files in `pkg/eval/vals`). I will do a followup commit to implement introspection in my preferred fashion.

Reply:
Without introspection the output of `path:temp-file` isn't really useful since there wouldn't be any way to delete the temporary file. I'll implement indexing support so `$file[fd]` and `$file[name]` yield useful values. I'll leave it to you to decide whether, and how, a `*os.File` is exposed as a pseudo-map so commands like `keys` work on such objects. I'll also squash the commits and rebase on the current master branch.